hoomd/trigger.py

```diff
@@ -1,14 +1,58 @@
 # Copyright (c) 2009-2019 The Regents of the University of Michigan
-# This file is part of the HOOMD-blue project, released under the BSD 3-Clause
 # License.
+# This file is part of the HOOMD-blue project, released under the BSD 3-Clause
 
 from hoomd import _hoomd
+from inspect import isclass
 
 
 class Trigger(_hoomd.Trigger):
     pass
 
 
-class PeriodicTrigger(_hoomd.PeriodicTrigger, Trigger):
+class Periodic(_hoomd.PeriodicTrigger, Trigger):
     def __init__(self, period, phase=0):
         _hoomd.PeriodicTrigger.__init__(self, period, phase)
+
+
+class Before(_hoomd.BeforeTrigger, Trigger):
+    def __init__(self, timestep):
+        if timestep < 0:
+            raise ValueError("timestep must be positive.")
+        else:
+            _hoomd.BeforeTrigger.__init__(self, timestep)
+
+
+class On(_hoomd.OnTrigger, Trigger):
+    def __init__(self, timestep):
+        if timestep < 0:
+            raise ValueError("timestep must be positive.")
+        else:
+            _hoomd.OnTrigger.__init__(self, timestep)
+
+
+class After(_hoomd.AfterTrigger, Trigger):
+    def __init__(self, timestep):
+        if timestep < 0:
+            raise ValueError("timestep must be positive.")
+        else:
+            _hoomd.AfterTrigger.__init__(self, timestep)
+
+
+class Not(_hoomd.NotTrigger, Trigger):
+    def __init__(self, trigger):
+        _hoomd.NotTrigger.__init__(self, trigger)
+
+
+class And(_hoomd.AndTrigger, Trigger):
+    def __init__(self, triggers):
+        if not hasattr(triggers, '__iter__') or isclass(triggers):
```

Comment:
You might have a generator of triggers that is a class. I think that wouldn't work with this check in place. Consider:
```python
from collections.abc import Iterable

if isinstance(e, Iterable):
    # e is iterable
```
https://stackoverflow.com/questions/1952464/in-python-how-do-i-determine-if-an-object-is-iterable

Reply:
You might try something like:
```python
class And(_hoomd.AndTrigger, Trigger):
    def __init__(self, triggers):
        triggers = list(triggers)  # resolves generators and other weird iterables into a list
        if not all(isinstance(t, Trigger) for t in triggers):  # ensure all items are triggers (skip this if the check can be done safely in C++!)
            raise ValueError("triggers must an iterable of Triggers.")
        _hoomd.AndTrigger.__init__(self, triggers)
```

Reply:
That seems clean.